File: editors/vscode/src/protocol.ts
import * as path from "node:path";

export interface LintDiagnostic {
  path: string;
  line: number;
  column: number;
  severity: "warning" | "error";
  rule: string;
  message: string;
  help?: string | null;
}

export interface LintReport {
  diagnostics: LintDiagnostic[];
  error_count: number;
  warning_count: number;
}

export interface BuildReport {
  pdf_path?: string | null;
  elapsed_ms: number;
  skipped: boolean;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

export function parseLintReport(stdout: string): LintReport {
  const value: unknown = JSON.parse(stdout);
  if (!isObject(value) || !Array.isArray(value.diagnostics)) {
    throw new Error("Tekai returned an invalid lint report");
  }
  const diagnostics = value.diagnostics.map((entry): LintDiagnostic => {
    if (
      !isObject(entry) ||
      typeof entry.path !== "string" ||
      typeof entry.line !== "number" ||
      typeof entry.column !== "number" ||
      (entry.severity !== "warning" && entry.severity !== "error") ||
      typeof entry.rule !== "string" ||
      typeof entry.message !== "string"
    ) {
      throw new Error("Tekai returned an invalid diagnostic");
    }
    return {
      path: entry.path,
      line: entry.line,
      column: entry.column,
      severity: entry.severity,
      rule: entry.rule,
      message: entry.message,
      help: typeof entry.help === "string" ? entry.help : null,
    };
  });
  return {
    diagnostics,
    error_count: typeof value.error_count === "number" ? value.error_count : 0,
    warning_count: typeof value.warning_count === "number" ? value.warning_count : 0,
  };
}

export function parseBuildReport(stdout: string): BuildReport {
  const value: unknown = JSON.parse(stdout);
  if (!isObject(value) || typeof value.elapsed_ms !== "number") {
    throw new Error("Tekai returned an invalid build report");
  }
  return {
    pdf_path: typeof value.pdf_path === "string" ? value.pdf_path : null,
    elapsed_ms: value.elapsed_ms,
    skipped: value.skipped === true,
  };
}

export function parseMagicRoot(source: string): string | undefined {
  const match = source
    .split(/\r?\n/, 50)
    .join("\n")
    .match(/^\s*%\s*!\s*TEX\s+root\s*=\s*(.+?)\s*$/im);
  if (!match) {
    return undefined;
  }
  return match[1].trim().replace(/^(["'])(.*)\1$/, "$2");
}

export function parseBuiltPdf(line: string): string | undefined {
  const match = line.match(/(?:built|cached)\s+(.+\.pdf)\s+in\s+[\d.]+(?:ns|us|µs|ms|s)/i);
  return match?.[1].trim();
}

export function absoluteReportedPath(reportedPath: string, cwd: string): string {
  return path.isAbsolute(reportedPath) ? path.normalize(reportedPath) : path.resolve(cwd, reportedPath);
}

File: editors/nvim/lua/tekai/protocol.lua
local M = {}

function M.magic_root(source)
  local count = 0
  for line in (source .. "\n"):gmatch("(.-)\r?\n") do
    count = count + 1
    if count > 50 then
      break
    end
    local root = line:match("^%s*%%%s*!%s*[Tt][Ee][Xx]%s+[Rr][Oo][Oo][Tt]%s*=%s*(.-)%s*$")
    if root and root ~= "" then
      local quote, value = root:match("^([\"'])(.*)%1$")
      return quote and value or root
    end
  end
  return nil
end

function M.decode_lint(stdout)
  local ok, report = pcall(vim.json.decode, stdout)
  if not ok or type(report) ~= "table" or type(report.diagnostics) ~= "table" then
    return nil, "Tekai returned an invalid lint report"
  end
  for _, diagnostic in ipairs(report.diagnostics) do
    if
      type(diagnostic) ~= "table"
      or type(diagnostic.path) ~= "string"
      or type(diagnostic.line) ~= "number"
      or type(diagnostic.column) ~= "number"
      or (diagnostic.severity ~= "warning" and diagnostic.severity ~= "error")
      or type(diagnostic.rule) ~= "string"
      or type(diagnostic.message) ~= "string"
    then
      return nil, "Tekai returned an invalid diagnostic"
    end
  end
  return report
end

function M.decode_build(stdout)
  local ok, report = pcall(vim.json.decode, stdout)
  if not ok or type(report) ~= "table" or type(report.elapsed_ms) ~= "number" then
    return nil, "Tekai returned an invalid build report"
  end
  return report
end

function M.decode_check(stdout)
  local report, err = M.decode_lint(stdout)
  if not report then
    return nil, err:gsub("lint report", "check report")
  end
  if report.elapsed_ms ~= nil and type(report.elapsed_ms) ~= "number" then
    return nil, "Tekai returned an invalid check build report"
  end
  if report.pdf_path ~= nil and type(report.pdf_path) ~= "string" then
    return nil, "Tekai returned an invalid check PDF path"
  end
  return report
end

function M.built_pdf(line)
  local pdf = line:match("[Bb][Uu][Ii][Ll][Tt]%s+(.+%.pdf)%s+in%s+[%d%.]+[%aµ]+")
    or line:match("[Cc][Aa][Cc][Hh][Ee][Dd]%s+(.+%.pdf)%s+in%s+[%d%.]+[%aµ]+")
  return pdf and vim.trim(pdf) or nil
end

return M
